### ISSUE TITLE
Constant propagation is not taking place for memcpy on struct with _Bool member

### ISSUE THREAD

**Issue by andreast271** [diffblue/cbmc#5542]
CBMC version: 5.17.0 (a46f12d66)
Operating system: Linux x86_64 (Red Hat 6.5)
Exact command line resulting in the issue: cbmc structcopy.c
What behaviour did you expect: constants are being propagated from source struct to destination struct 
What happened instead: constants are not being propagated

Example file:
[structcopy.c.txt](https://github.com/diffblue/cbmc/files/5484064/structcopy.c.txt)

Constant propagation is not taking place in the following function call:
`memcpy((char*)&my_conf, (char*)&conf_init, 1 * sizeof(struct conft));`

The source of the memcpy operation is a struct populated with constants. If the struct has only members of type char, short, int, then constant propagation does take place. However, if the struct has an element of type _Bool, constant propagation does not take place.

**Comment by martin-cs** [comment]
Hmmmm... that sounds like a bug.  Thanks for the really clear test case.  I think the problem is in constant folding `byte_extract`:

```
$ cbmc memcpy-struct-bug.c --unwind 1 --show-vcc --slice-formula
CBMC version 5.17.0 (cbmc-5.17.0-353-g65a785424-dirty) 64-bit x86_64 linux
Parsing memcpy-struct-bug.c
Converting
Type-checking memcpy-struct-bug
file memcpy-struct-bug.c line 34 function main: function 'assert' is not declared
Generating GOTO Program
Adding CPROVER library (x86_64)
Removal of function pointers and virtual functions
Generic Property Instrumentation
Running with 8 object bits, 56 offset bits (default)
Starting Bounded Model Checking
Not unwinding loop main.0 iteration 1 file memcpy-struct-bug.c line 35 function main thread 0
size of program expression: 89 steps
slicing removed 35 assignments
Generated 4 VCC(s), 1 remaining after simplification
VERIFICATION CONDITIONS:

file memcpy-struct-bug.c line 34 function main
assertion num == 1
{-1} memcpy::1::1::src_n!0@1#1 = byte_extract_little_endian({ 0, 0, 1 }, 0, signedbv[8][8])
{-2} memcpy::1::1::src_n!0@1#2[[0]] = memcpy::1::1::src_n!0@1#1[0]
{-3} memcpy::1::1::src_n!0@1#2[[1]] = memcpy::1::1::src_n!0@1#1[1]
{-4} memcpy::1::1::src_n!0@1#2[[2]] = memcpy::1::1::src_n!0@1#1[2]
{-5} memcpy::1::1::src_n!0@1#2[[3]] = memcpy::1::1::src_n!0@1#1[3]
{-6} memcpy::1::1::src_n!0@1#2[[4]] = memcpy::1::1::src_n!0@1#1[4]
{-7} memcpy::1::1::src_n!0@1#2[[5]] = memcpy::1::1::src_n!0@1#1[5]
{-8} memcpy::1::1::src_n!0@1#2[[6]] = memcpy::1::1::src_n!0@1#1[6]
{-9} memcpy::1::1::src_n!0@1#2[[7]] = memcpy::1::1::src_n!0@1#1[7]
{-10} main::1::my_conf!0@1#1 = byte_extract_little_endian({ memcpy::1::1::src_n!0@1#2[[0]], memcpy::1::1::src_n!0@1#2[[1]], memcpy::1::1::src_n!0@1#2[[2]], memcpy::1::1::src_n!0@1#2[[3]], memcpy::1::1::src_n!0@1#2[[4]], memcpy::1::1::src_n!0@1#2[[5]], memcpy::1::1::src_n!0@1#2[[6]], memcpy::1::1::src_n!0@1#2[[7]] }, 0, struct tag-conft)
{-11} main::1::my_conf!0@1#2..one = main::1::my_conf!0@1#1.one
{-12} main::1::num!0@1#2 = main::1::my_conf!0@1#2..one
├──────────────────────────
{1} main::1::num!0@1#2 = 1
```

**Comment by martin-cs** [comment]
I think the problem is that `ID_c_bool` (and maybe `ID_bool`) is not handled in `simplify_exprt::expr2bits` in this if/else chain:

https://github.com/diffblue/cbmc/blob/fa4569e1bfbb0ee3b6b4f54e6b8aea4d3911c6ea/src/util/simplify_expr.cpp#L1775

adding it should be reasonably straight-forward.  I may try to create a PR next week if I have time.

**Comment by martin-cs** [comment]
This is definitely part of the solution but not the complete fix.

**Comment by tautschnig** [comment]
@martin-cs Thank you for diagnosing the problem! I've posted the fix in #5754. @andreast271 thank you for this nice test, I've included that in the PR.
